From 1c724d854e0f46c13690fb2fc1b98b80f80ec310 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Sat, 6 Nov 2021 00:00:56 +0100
Subject: [PATCH 0307/1695] plat-stm32mp1: chip and platform identification

New platform function to get the chip identification using
DBGMCU SoC register.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git a42cca1fba1264e88a0009e702be0da39703d185 plat-stm32mp1: chip and platform identification]
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Change-Id: I56727a701ded07e849c17fdf06e156ee3b53d0c7
---
 core/arch/arm/plat-stm32mp1/main.c            | 47 +++++++++++++++++++
 core/arch/arm/plat-stm32mp1/platform_config.h |  6 +++
 core/arch/arm/plat-stm32mp1/stm32_util.h      |  4 ++
 3 files changed, 57 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index dbbd48027..373eec118 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -147,6 +147,53 @@ static TEE_Result init_console_from_dt(void)
 /* Probe console from DT once clock inits (service init level) are completed */
 service_init_late(init_console_from_dt);
 
+static uintptr_t stm32_dbgmcu_base(void)
+{
+	static void *va;
+
+	if (!cpu_mmu_enabled())
+		return DBGMCU_BASE;
+
+	if (!va)
+		va = phys_to_virt(DBGMCU_BASE, MEM_AREA_IO_SEC, 1);
+
+	return (uintptr_t)va;
+}
+
+/* SoC versioning util, returns default ID if can't access DBGMCU */
+TEE_Result stm32mp1_dbgmcu_get_chip_version(uint32_t *chip_version)
+{
+	uint32_t id = STM32MP1_CHIP_DEFAULT_VERSION;
+
+	if (!chip_version)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (stm32_bsec_read_debug_conf() & BSEC_DBGSWGEN)
+		id = (io_read32(stm32_dbgmcu_base() + DBGMCU_IDC) &
+		      DBGMCU_IDC_REV_ID_MASK) >> DBGMCU_IDC_REV_ID_SHIFT;
+
+	*chip_version = id;
+
+	return TEE_SUCCESS;
+}
+
+/* SoC device ID util, returns default ID if can't access DBGMCU */
+TEE_Result stm32mp1_dbgmcu_get_chip_dev_id(uint32_t *chip_dev_id)
+{
+	uint32_t id = STM32MP1_CHIP_ID;
+
+	if (!chip_dev_id)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (stm32_bsec_read_debug_conf() & BSEC_DBGSWGEN)
+		id = io_read32(stm32_dbgmcu_base() + DBGMCU_IDC) &
+		     DBGMCU_IDC_DEV_ID_MASK;
+
+	*chip_dev_id = id;
+
+	return TEE_SUCCESS;
+}
+
 /*
  * GIC init, used also for primary/secondary boot core wake completion
  */
diff --git a/core/arch/arm/plat-stm32mp1/platform_config.h b/core/arch/arm/plat-stm32mp1/platform_config.h
index 772068e6b..12152d9cd 100644
--- a/core/arch/arm/plat-stm32mp1/platform_config.h
+++ b/core/arch/arm/plat-stm32mp1/platform_config.h
@@ -258,6 +258,12 @@
 #define USART3_BASE			UART3_BASE
 #define USART6_BASE			UART6_BASE
 
+/* DBGMCU resources */
+#define DBGMCU_IDC			U(0x0)
+#define DBGMCU_IDC_DEV_ID_MASK		GENMASK_32(11, 0)
+#define DBGMCU_IDC_REV_ID_MASK		GENMASK_32(31, 16)
+#define DBGMCU_IDC_REV_ID_SHIFT		U(16)
+
 /* SYSRAM layout */
 #ifdef CFG_STM32MP13
 #define SYSRAM_SIZE			0x20000
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index 1a5ba6451..9dbc16613 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -16,6 +16,10 @@
 #include <tee_api_types.h>
 #include <types_ext.h>
 
+/* SoC versioning and device ID */
+TEE_Result stm32mp1_dbgmcu_get_chip_version(uint32_t *chip_version);
+TEE_Result stm32mp1_dbgmcu_get_chip_dev_id(uint32_t *chip_dev_id);
+
 /* Backup registers and RAM utils */
 vaddr_t stm32mp_bkpreg(unsigned int idx);
 
-- 
2.47.2

